
# GWAS Data Quality Control (QC)

QC procedures and statistical analyses will be illustrated using the free, open‐source whole‐genome association analysis toolset PLINK version 1.9 (<https://www.cog-genomics.org/plink/1.9/>). The newer version of PLINK (PLINK 2) is also available and it's possible to complete the QC steps using PLINK 2. Though, keep in mind instead of bfiles (for PLINK 1.9) you'll need to generate **.gen** files for PLINK 2. ). I'll be using R packages to isuaize the outcomes at each QC step. 

You may benefit studying these tutorials as well:  <https://plink.readthedocs.io/en/latest/plink_fmt/>
<https://kcniconfluence.camh.ca/display/GEN/Genome+Wide+Quality+Control+and+Basic+Association+Analyses>
<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6001694/>
<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3066182/>
<https://github.com/choishingwan/PRS-Tutorial/blob/master/docs/base.md> 

Most of the QC steps involve calculating summaries at the (a) individual ("row") level or the (b) SNP ("column") level. 

(a) *For individuals*, you'll need to check on 1) missing data, 2) relatedness, and 3) sex chromosome dosage. You'll need to exclude those with high missingness and those genetically related. Also, if the individuals registered as male or female in the metadata do not match the sex estimation based on the allelic dose for X chromosome (mismatches). 

(b) *For variants (SNPs)*, check on the following QC measures: 

     b.1) Chromosome check: This is not exactly a QC step, but extremely  
          helpful to do as a first step when getting any genetic data, it's 
          always good to check which chromosome the SNPs are located on 
          (autosomal, 1-22; X; Y; MT; or 0=SNPs can not  be mapped to any 
          chromosome). 
          
     b.2) Missing/ duplicated data: a common practice is to exclude data with 
          more than 5% missing data.
          
     b.3) Minor Allel Frequency (MAF): SNPs with low minor allele frequency
          (MAF) are more likely to generate false positive results due to their
          lower statistical power (and higher probability of genotyping errors 
          in the case of low MAF). Therefore,usually alleles with MAF <0.05 
          (sometimes <0.01) are deleted.  
           
          
     b.4) Hardy-Weinberg Equilibrium (HWE): exclude the SNPs only if the HWE is
          hugely violated (e.g., p-value<10-10 for a test of whether the data 
          follow a binomial distribution). The HW prinicple states that alleles
          frequencies in a population remains constant from generation to  
          generation in the absence of other evolutionary forces. The 
          distribution of genotypes should folow a binomial distribution with 
          probability  equal to the MAF. If this doesn't happen, it means 
          either the genotyping for that SNP has error or the Mating was not 
          random. 
          
### QC Steps

#### Step 1) Check discordant sex information
Subjects who were a priori determined as females must have a F value of <0.2, and subjects who were a priori determined as males must have a F value >0.8. This F value is based on the X chromosome inbreeding (homozygosity) estimate.
Subjects who do not fulfil these requirements are flagged "PROBLEM" by PLINK.

```{bash}

$ plink --memory 128000 --bfile gtex_wgs_white --check-sex --out sample_sex

```

```{bash}

# Generate plots to visualize the sex-check results.
$ Rscript --no-save gender_check.R

```

This will generate a bar plot showing the number of subjects with Sex discrepancies.

The following two scripts can be used to deal with individuals with a sex discrepancy. Note, please use one of the two options below to generate a new bfile, this file we will use in the next steps of this tutorial.

```{bash}

# 1) Delete individuals with sex discrepancy.
$ grep "PROBLEM" plink.sexcheck| awk '{print$1,$2}'> sex_discrepancy.txt
# This command generates a list of individuals with the status “PROBLEM”.
$ plink --bfile gtex_wgs_white --remove sex_discrepancy.txt --make-bed --out gtex_wgs_white_1 
# This command removes the list of individuals with the status “PROBLEM”.


# or

# 2) impute-sex.
$ plink --bfile gtex_wgs_white --impute-sex --make-bed --out gtex_wgs_white_1
# This imputes the sex based on the genotype information into your data set.

```


#### Step 2) Investigate missingness per individual and per SNP and make histograms.

```{bash}

#!/bin/bash

#SBATCH --nodes=1
#SBATCH --cpus-per-task=12
#SBATCH --mem-per-cpu=128000M
#SBATCH --time=12:00:00
#SBATCH --job-name="GTeX_QC"
#SBATCH --output=/external/rprshnas01/netdata_kcni/stlab/Nickie/GTEX_V8/Genotype_Data/3_QCed_Data/gtex_QC.out
#SBATCH --error=/external/rprshnas01/netdata_kcni/stlab/Nickie/GTEX_V8/Genotype_Data/3_QCed_Data/gtex_QC.err       

$ plink --memory 128000 --bfile gtex_wgs_white_1 --missing 

# output: plink.imiss and plink.lmiss, these files show respectively the proportion of missing SNPs per individual and the proportion of missing individuals per SNP.

```


```{bash}

# Generate plots to visualize the missingness results.
Rscript --no-save hist_miss.R

```


```{bash}

# Delete SNPs and individuals with high levels of missingness.
# The following two QC commands will not remove any SNPs or individuals. However, it is good practice to start the QC with these non-stringent thresholds.  
# Delete SNPs with missingness >0.2.
plink --memory 128000 --bfile gtex_wgs_white_1 --geno 0.2 --make-bed --out gtex_wgs_white_2

# Delete individuals with missingness >0.2.
plink --memory 128000 --bfile gtex_wgs_white_2 --mind 0.2 --make-bed --out gtex_wgs_white_3

# Delete SNPs with missingness >0.02.
plink --memory 128000 --bfile gtex_wgs_white_3 --geno 0.02 --make-bed --out gtex_wgs_white_4

# Delete individuals with missingness >0.02.
plink --memory 128000 --bfile gtex_wgs_white_4 --mind 0.02 --make-bed --out gtex_wgs_white_5

```


#### Step 3) Generate a bfile with autosomal SNPs only and delete SNPs with a low minor allele frequency (MAF).

```{bash}

# Select autosomal SNPs only (i.e., from chromosomes 1 to 22).
$ awk '{ if ($1 >= 1 && $1 <= 22) print $2 }' gtex_wgs_white_5.bim > snp_1_22.txt

$ plink --bfile gtex_wgs_white_5 --extract snp_1_22.txt --make-bed --out gtex_wgs_white_6

```

```{bash}

# Generate a plot of the MAF distribution.
$ plink --bfile gtex_wgs_white_6 --freq --out MAF_check

$ Rscript --no-save MAF_check.R

```


```{bash}

# Remove SNPs with a low MAF frequency.
$ plink --memory 128000 --bfile gtex_wgs_white_6 --maf 0.05 --make-bed --out gtex_wgs_white_7

# 1073226 SNPs are left
# A conventional MAF threshold for a regular GWAS is between 0.01 or 0.05, depending on sample size.

```


#### Step 4) Delete SNPs which are not in Hardy-Weinberg equilibrium (HWE).

```{bash}


# Check the distribution of HWE p-values of all SNPs.
$ plink --memory 128000 --bfile gtex_wgs_white_7 --hardy

# Selecting SNPs with HWE p-value below 0.00001, required for one of the two plot generated by the next Rscript, allows to zoom in on strongly deviating SNPs. 

$ awk '{ if ($9 <0.00001) print $0 }' plink.hwe>plinkzoomhwe.hwe
$ Rscript --no-save hwe.R

# By default the --hwe option in plink only filters for controls.
# Therefore, we use two steps, first we use a stringent HWE threshold for controls, followed by a less stringent threshold for the case data.

$ plink --memory 128000 --bfile gtex_wgs_white_7 --hwe 1e-6 --make-bed --out gtex_wgs_white_7_hwe_filter_step1

# The HWE threshold for the cases filters out only SNPs which deviate extremely from HWE. 
# This second HWE step only focusses on cases because in the controls all SNPs with a HWE p-value < hwe 1e-6 were already removed
$ plink --memory 128000 --bfile gtex_wgs_white_7_hwe_filter_step1 --hwe 1e-10 --hwe-all --make-bed --out gtex_wgs_white_8


```


#### Step 5) Generate a plot of the distribution of the heterozygosity rate of your subjects. And remove individuals with a heterozygosity rate deviating more than 3 sd from the mean.

```{bash}

# Checks for heterozygosity are performed on a set of SNPs which are not highly correlated. Therefore, to generate a list of non-(highly)correlated SNPs, we exclude high inversion regions (inversion.txt [High LD regions]) and prune the SNPs using the command --indep-pairwise’.

# The parameters ‘50 5 0.2’ stand respectively for: the window size, the number of SNPs to shift the window at each step, and the multiple correlation coefficient for a SNP being regressed on all other SNPs simultaneously.

$ plink --memory 128000 --bfile gtex_wgs_white_8 --exclude inversion.txt --range --indep-pairwise 50 5 0.2 --out indepSNP

# Note, don't delete the file indepSNP.prune.in, we will use this file in later steps of the tutorial.

$ plink --memory 128000 --bfile gtex_wgs_white_8 --extract indepSNP.prune.in --het --out R_check
# This file contains your pruned data set.

# Plot of the heterozygosity rate distribution
$ Rscript --no-save check_heterozygosity_rate.R

# The following code generates a list of individuals who deviate more than 3 standard deviations from the heterozygosity rate mean.
# For data manipulation we recommend using UNIX. However, when performing statistical calculations R might be more convenient, hence the use of the Rscript for this step:
$ Rscript --no-save heterozygosity_outliers_list.R

# Output of the command above: fail-het-qc.txt .
# When using our example data/the GTex data this list contains 4 individuals (i.e., two individuals have a heterozygosity rate deviating more than 3 SD's from the mean).
# Adapt this file to make it compatible for PLINK, by removing all quotation marks from the file and selecting only the first two columns.

$ sed 's/"// g' fail-het-qc.txt | awk '{print$1, $2}'> het_fail_ind.txt

# Remove heterozygosity rate outliers.
$ plink --memory 128000 --bfile gtex_wgs_white_8 --remove het_fail_ind.txt --make-bed --out gtex_wgs_white_9


```


#### Step 6) Check datasets for cryptic relatedness.
Assuming a random population sample we are going to exclude all individuals 
above the pihat threshold of 0.2 in this tutorial.

```{bash}

# Check for relationships between individuals with a pihat > 0.2.
$ plink --memory 128000 --bfile gtex_wgs_white_9 --extract indepSNP.prune.in --genome --min 0.2 --out pihat_min0.2

# I am not sure if The GTEX dataset contain parent-offspring relations. 
# The following commands will visualize specifically the parent-offspring relations, using the z values. 

$ awk '{ if ($8 >0.9) print $0 }' pihat_min0.2.genome>zoom_pihat.genome

# Generate a plot to assess the type of relationship.
$ Rscript --no-save Relatedness.R

# The generated plots show amount of related individuals (explentation plot; PO = parent-offspring, UN = unrelated individuals). 
# Normally, family based data should be analyzed using specific family based methods. In this tutorial, for demonstrative purposes, we treat the relatedness as cryptic relatedness in a random population sample.


# In this tutorial, we aim to remove all 'relatedness' from our dataset.
# To demonstrate that the majority of the relatedness was due to parent-offspring we only include founders (individuals without parents in the dataset).

$ plink --bfile gtex_wgs_white_9 --filter-founders --make-bed --out gtex_wgs_white_10

# Now we will look again for individuals with a pihat >0.2.
plink --bfile gtex_wgs_white_10 --extract indepSNP.prune.in --genome --min 0.2 --out pihat_min0.2_in_founders

# The file 'pihat_min0.2_in_founders.genome' shows that, after exclusion of all non-founders, only 1 individual pair with a pihat greater than 0.2 remains in the HapMap data.

# This is likely to be a full sib or DZ twin pair based on the Z values. Noteworthy, they were not given the same family identity (FID) in the HapMap data.

# For each pair of 'related' individuals with a pihat > 0.2, we recommend to remove the individual with the lowest call rate. 
plink --bfile gtex_wgs_white_10 --missing
# Use an UNIX text editor (e.g., vi(m) ) to check which individual has the highest call rate in the 'related pair'. 

# Generate a list of FID and IID of the individual(s) with a Pihat above 0.2, to check who had the lower call rate of the pair.
# In our dataset the individual 13291  NA07045 had the lower call rate.
vi 0.2_low_call_rate_pihat.txt

# Press esc on keyboard!
:x
# Press enter on keyboard
# In case of multiple 'related' pairs, the list generated above can be extended using the same method as for our lone 'related' pair.

# Delete the individuals with the lowest call rate in 'related' pairs with a pihat > 0.2 
plink --bfile gtex_wgs_white_10 --remove 0.2_low_call_rate_pihat.txt --make-bed --out gtex_wgs_white_11



```

Congratulations! The QC procedure is completed!


